### ISSUE TITLE
[InstCombine] Missed optimization : fold `binop(select cond, C1, C2, zext(cond))`

### ISSUE THREAD

**Issue by XChy** [llvm/llvm-project#86176]
Alive2 proof: https://alive2.llvm.org/ce/z/-f5J6T

### Motivating example 

```llvm
define i32 @src(i1 %c) {
entry:
  %cond_zext = zext i1 %c to i32
  %s = select i1 %c, i32 67108863, i32 0
  %or = or i32 %s, %cond_zext
  ret i32 %or
}
```
can be folded to:
```llvm
define i32 @tgt(i1 %c) {
entry:
  %s = select i1 %c, i32 67108863, i32 0
  ret i32 %s
}
```

### Real-world motivation

This snippet of IR is derived from [qemu/system/physmem.c@phys_page_set_level](https://github.com/qemu/qemu/blob/fea445e8fe9acea4f775a832815ee22bdf2b0222/system/physmem.c#L204) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, please email me.

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

**Comment by gwastar** [comment]
This fold already happens for add, sub, and mul and was introduced in commit  f12a556. It is performed in foldBinOpOfSelectAndCastOfSelectCondition. This function seems to have two issues: It doesn't preserve flags like nuw, nsw, or disjoint and it allows multi-use selects which I don't think is profitable (see https://godbolt.org/z/vo8GY6xsa and https://alive2.llvm.org/ce/z/AtLvvW).  I guess I should fix that  function first so that I can use it for this.

**Comment by XChy** [comment]
@gwastar, yes, I looked into this function. One-use constraint is needed for the select at least.
You could first file a patch to add some constraints on it, and then file the second patch for this issue.
